text/1006-deprecate-action-template-helper.md

```diff
@@ -0,0 +1,205 @@
+project-link: Leave as is
+-->
+
+# Deprecate `(action)` template helper and `{{action}}` modifier. 
+
+## Summary
+
+The `(action)` template helper and `{{action}}` modifier was common pre-Octane. Now that we have native classes and the `{{on}}` modifier, we no longer need to use `(action)` or `{{action}}`
+
+## Motivation
+
+Remove legacy code with confusing semantics.
+
+This is a part of _[Deprecating Ember Classic (pre-Octane)](https://github.com/emberjs/rfcs/issues/832)_.
+
+## Transition Path
+
+This was written in the [Octave vs Classic cheatsheet](https://ember-learn.github.io/ember-octane-vs-classic-cheat-sheet/#component-properties__ddau)
+
+<details><summary>that content here</summary>
+
+### Before (pre-Octane)
+
+```js
+// parent-component.js
+import Component from '@ember/component';
+
+export default Component.extend({
+  count: 0
+});
+
+```
+```hbs
+{{!-- parent-component.hbs --}}
+{{child-component count=count}}
+Count: {{this.count}}
+
+```
+```js
+// child-component.js
+import Component from '@ember/component';
+
+export default Component.extend({
+  actions: {
+    plusOne() {
+      this.set('count', this.get('count') + 1);
+    }
+  }
+});
+```
+```hbs
+{{!-- child-component.hbs --}}
+<button type="button" {{action "plusOne"}}>
+  Click Me
+</button>
+```
+
+### After (post-Octane)
+```js
+// parent-component.js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+
+export default class ParentComponent extends Component {
+  @tracked count = 0;
+
+  @action plusOne() {
+    this.count++;
+  }
+}
+
+```
+```hbs
+{{!-- parent-component.hbs --}}
+<ChildComponent @plusOne={{this.plusOne}} />
+Count: {{this.count}}
+
+```
+```hbs
+{{!-- child-component.hbs --}}
+<button type="button" {{on "click" @plusOne}}>
+  Click Me
+</button>
+
+```
+
+</details>
+
+But what we could put in the deprecation app:
+
+### Scenario: `action` is passed a string
+
+Before:
+```hbs
+<button type="button" {{action "plusOne"}}>
+  Click Me
+</button>
+```
+
+After
+
+```hbs
+<button type="button" {{on 'click' this.plusOne}}>
+  Click Me
+</button>
+```
+or, if `plusOne` is passed in as an argument 
+```hbs
+<button type="button" {{on 'click' @plusOne}}>
+  Click Me
+</button>
+```
+
+### Scenario: `action` is passed a function reference
+
+Before:
+```hbs
+<SomeComponent @update={{action this.plusOne}} />
+```
+
+After
+
+```hbs
+<SomeComponent @update={{this.plusOne}} />
+```
+
+### Scenario: `action` is passed parameters
+
+Before:
+```hbs
+<SomeComponent @update={{action this.plus 1}} />
+```
+
+After:
+```hbs
+<SomeComponent @update={{fn this.plus 1}} />
+```
+
+### Scenario: `action` is used with `mut` 
+
+Before:
+```hbs
+<SomeComponent @update={{action (mut @value.property}} />
+```
+After:
+```js
+// parent.js
+export default class SomeComponent extends Component {
+    handleUpdate = (value) => this.args.property = value; 
```

Comment:
Have we recommended this style? I thought we were still recommending the `@action` decorator.

Reply:
`@action` is mostly a back-compat thing to allow for templates to keep using `(action)` / `{{action}}` -- beyond binding a normal method (rather than a property with a value of function) I don't think it serves any purpose. Happy to be wrong, but :Shrug:

Reply:
I tried dropping `@action` at one point and replacing it with that style. I was blocked in a lot of cases by overridden actions that called the parent class's action with `super`. Is there a good replacement for that?

Reply:
class properties can't override a thing-of-the-same-name, so probably not.
maybe `@action` does stick around for this reason (not mentioned here in this RFC) -- after `(action)` and `{{action}}` are removed (v6 🤞), we can remove all the compatibility code in `@action` and have it _just_ be about binding methods.

Reply:
afaik `@action` just binds, but I think we had intentionally chosen that over the arrow function and teach that in the guides, iirc.

Reply:
yeah, a discussion for another day -- I've updated this example to use the action decorator

Reply:
If it needs to stick around just for the binding effect, I'd suggest creating a new `@bind` or `@bound` decorator as a solution to move to, while then deprecating `@action`. I'd rather not have to explain to someone new to ember that "Well.... `@action` is an old ember thing but we still use that just to bind a method". Rather, just make a new decorator that has a clear intent. My 2 cents anyway.

Reply:
`@action` was only added with the move to native classes, so it really isn't an old ember thing. The `actions` hash and the `action` modifier _are_ old ember things.

Reply:
Right, but the term comes from an old emberism. Still would be confusing for anyone who doesn't know the history of actions.

Reply:
Or is it less confusing because the only "action" they know is `@action`?

Reply:
No, because it's still not clear what `@action` does, you have to teach what it does. Whereas `@bind` is fairly clear what you would expect it to do, provided you know about `.bind()` I suppose.